File: typescript/tests/trials/interop.test.ts
/**
 * Cross-port hash parity. Both ports load fixtures/trials/canonical.json and
 * MUST produce the same SHA-256 hex. If either side drifts the interop CI
 * matrix catches it. The Go counterpart lives at
 * licensing/trials/interop_test.go.
 */

import { describe, expect, it } from 'bun:test';
import { readFileSync } from 'node:fs';
import { join } from 'node:path';

import { hashFingerprint } from '../../src/trials/pepper.ts';

interface CanonicalFixture {
  readonly pepper: string;
  readonly fingerprint_input: string;
  readonly fingerprint_hash: string;
}

const FIXTURES_PATH = join(import.meta.dir, '../../../fixtures/trials/canonical.json');

describe('trial-pepper cross-port hash parity', () => {
  it('hashFingerprint matches the canonical fixture', () => {
    const fx = JSON.parse(readFileSync(FIXTURES_PATH, 'utf8')) as CanonicalFixture;
    const got = hashFingerprint(fx.pepper, fx.fingerprint_input);
    expect(got).toBe(fx.fingerprint_hash);
  });
});

File: licensing/trials/interop_test.go
package trials_test

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"testing"

	"github.com/AnoRebel/licensing/licensing/trials"
)

type canonicalFixture struct {
	Pepper           string `json:"pepper"`
	FingerprintInput string `json:"fingerprint_input"`
	FingerprintHash  string `json:"fingerprint_hash"`
}

func loadCanonical(t *testing.T) canonicalFixture {
	t.Helper()
	_, file, _, _ := runtime.Caller(0)
	root := filepath.Join(filepath.Dir(file), "..", "..")
	path := filepath.Join(root, "fixtures", "trials", "canonical.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	var fx canonicalFixture
	if err := json.Unmarshal(data, &fx); err != nil {
		t.Fatalf("unmarshal fixture: %v", err)
	}
	return fx
}

// TestCrossPortHashParity asserts the Go hash matches the canonical fixture
// produced by both ports. The TS side has the same test against the same
// fixture; if either port drifts, the cross-language interop suite catches it.
func TestCrossPortHashParity(t *testing.T) {
	fx := loadCanonical(t)
	got, err := trials.HashFingerprint(fx.Pepper, fx.FingerprintInput)
	if err != nil {
		t.Fatal(err)
	}
	if got != fx.FingerprintHash {
		t.Fatalf("hash mismatch:\n  got:  %s\n  want: %s", got, fx.FingerprintHash)
	}
}
